### PR TITLE
Python bindings

### PR DESCRIPTION
**Pull request by leo-stan**

### CODE REVIEW

File: python/bindings.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        .def("ExtendedScanAngle", static_cast<void (las::Point::*)(const int16_t &)>(&las::Point::ExtendedScanAngle))
+        .def("ScanAngle", static_cast<float (las::Point::*)() const>(&las::Point::ScanAngle))
+        .def("ScanAngle", static_cast<void (las::Point::*)(const float &)>(&las::Point::ScanAngle))
+        .def("UserData", static_cast<uint8_t (las::Point::*)() const>(&las::Point::UserData))
+        .def("UserData", static_cast<void (las::Point::*)(const uint8_t &)>(&las::Point::UserData))
+        .def("PointSourceID", static_cast<uint16_t (las::Point::*)() const>(&las::Point::PointSourceID))
+        .def("PointSourceID", static_cast<void (las::Point::*)(const uint16_t &)>(&las::Point::PointSourceID))
+        .def("RGB",
+             static_cast<void (las::Point::*)(const uint16_t &, const uint16_t &, const uint16_t &)>(&las::Point::RGB))
+        .def("Red", static_cast<uint16_t (las::Point::*)() const>(&las::Point::Red))
```
Comment (CCInc) [inline python/bindings.cpp:98]: remember to update these to use properties ( https://pybind11.readthedocs.io/en/stable/classes.html#instance-and-static-fields , see the part where they talk about `def_property`)

File: CMakeLists.txt [hunk 2]
```diff
@@ -5,6 +5,7 @@ find_package(lazperf REQUIRED)
 
 set (COPC_SHARED_LIB copc-lib)
 set (COPC_STATIC_LIB copc-lib-s)
+set (COPC_PYTHON_LIB copclib)
```
[the remark this answers is not in the document]
Comment (CCInc) [reply CMakeLists.txt:8]: Or is this the name that we `import` into python?

File: test/points_test.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    point_list = [point1, copclib.Point(3, 4), copclib.Point(3, 4)]
+
+    points = copclib.Points(point_list)
+    assert points.PointFormatID() == 3
+    assert points.PointRecordLength() == 38
+    for point in points.Get():
+        assert point.PointFormatID() == 3
+    assert points.PointRecordLength() == 38
+    assert points.Get()[0].X() == 11
```
Comment (CCInc) [inline test/points_test.py:24]: I updated these in the C++ version to be `.Get(0)`

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (leo-stan) [reply CMakeLists.txt]: I've tried to change it but no success so far, it seems to be tied to the import name afaik

File: example/example-reader.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import copclib
+
+
+if __name__ == "__main__":
+
+    # Create a reader object
```
Comment (CCInc) [inline example/example-reader.py:6]: define ExampleReader() function or something then call it here

File: example/example-reader.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    reader = copclib.FileReader("../build/test/data/autzen-classified.copc.laz")
```
Comment (CCInc) [inline example/example-reader.py:7]: what are your thoughts on changing this to "copc" or something?

File: example/example-reader.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    print("WKT: %s\n" % reader.GetWkt())
+
+    load_key = copclib.VoxelKey(4, 11, 9, 0)
+
+    # FindNode will automatically load the minimum pages needed
+    # to find the key you request
+    node = reader.FindNode(load_key)
+    # If FindNode can't find the node, it will return an "invalid" node:
+    if not node.IsValid():
+        quit()
```
Comment (CCInc) [inline example/example-reader.py:34]: return?

File: example/example-writer.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+        # las::LasHeader header = writer.GetLasHeader()
+        # std::vector<char> uncompressed_points = reader.GetPointData(node)
+        # std::vector<char> compressed_points =
+        # laz::Compressor(uncompressed_points, header.point_format_id,
+        #                 cfg.extra_bytes.size(), header.point_record_length)
+        # writer.AddNodeCompressed(root_page, node.key, compressed_points, node.point_count)
+
+    # Make sure we call close to finish writing the file!
+    writer.Close()
```
Comment (CCInc) [inline example/example-writer.py:50]: remove?

File: test/writer_test.py [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+    writer.GetRootPage()
+    root_page = writer.GetRootPage()
+    assert root_page.IsValid()
+    assert root_page.IsPage()
+    assert root_page.loaded is True
+    with pytest.raises(RuntimeError):
+        writer.AddSubPage(root_page, copc.VoxelKey.InvalidKey())
+
+    writer.Close()
```
Comment (CCInc) [inline test/writer_test.py:107]: scoped

**Review by CCInc — COMMENTED** [review]
Looks good!

File: example/example-reader.py (hunk 5 above)
Comment (leo-stan) [reply example/example-reader.py:7]: I like it as `import copclib as copc` yep, will update to that

File: example/example-reader.py (hunk 6 above)
Comment (leo-stan) [reply example/example-reader.py:34]: Once it's defined as `ExampleReader()` function I will change to return yep

File: test/writer_test.py (hunk 8 above)
Comment (leo-stan) [reply test/writer_test.py:107]: Not sure what you mean here?

Comment (CCInc) [reply test/writer_test.py:107]: remove the Close and make the writer scoped

File: test/key_test.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    assert len(copc.VoxelKey(-1, -1, -1, -1).GetParents(False)) == 0
+
+    test_keys = [
+        copc.VoxelKey(4, 4, 6, 12),
+        copc.VoxelKey(3, 2, 3, 6),
+        copc.VoxelKey(2, 1, 1, 3),
+        copc.VoxelKey(1, 0, 0, 1),
+        copc.VoxelKey(0, 0, 0, 0),
+    ]
+    #
```
Comment (arknoll) [inline test/key_test.py:61]: Is this hash mark necessary?

Comment (leo-stan) [reply test/key_test.py:61]: Nop, I've just removed it

File: test/reader_test.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    # GetLasHeader Test
+    header = reader.GetLasHeader()
+    assert header.header_size == 375
+    assert header.point_format_id == 3
+    assert header.point_count == 10653336
+
+    # WKT Test
+    wkt = reader.GetWkt()
+    assert len(wkt) > 0
+    assert wkt.rfind("COMPD_CS[", 0) == 0
```
Comment (CCInc) [inline test/reader_test.py:30]:
```suggestion
    assert wkt.startswith("COMPD_CS[")
```

File: test/reader_test.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+    assert len(reader.GetPointData(key=invalid_node.key)) == 0
+    assert not len(reader.GetPointData(valid_node.key)) == 0
+
+    with pytest.raises(RuntimeError):
+        reader.GetPoints(invalid_node)
+    reader.GetPoints(valid_node)
+
+    assert len(reader.GetPoints(invalid_node.key).Get()) == 0
+    assert not len(reader.GetPoints(valid_node.key).Get()) == 0
```
Comment (CCInc) [inline test/reader_test.py:91]:
```suggestion
    assert len(reader.GetPoints(valid_node.key).Get()) != 0
```

File: test/reader_test.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+    reader = copc.FileReader("data/autzen-classified.copc.laz")
+
+    invalid_node = copc.Node()
+    with pytest.raises(RuntimeError):
+        reader.GetPointData(node=invalid_node)
+    valid_node = reader.FindNode(copc.VoxelKey(5, 9, 7, 0))
+    reader.GetPointData(valid_node)
+
+    assert len(reader.GetPointData(key=invalid_node.key)) == 0
+    assert not len(reader.GetPointData(valid_node.key)) == 0
```
Comment (CCInc) [inline test/reader_test.py:84]:
```suggestion
    assert len(reader.GetPointData(valid_node.key)) != 0
```

Comment (CCInc) [reply test/reader_test.py:84]: if we're going to use `==` for numerical conversions we may as well use `!=` as well

File: test/writer_test.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+    file_path = "data/writer_test.copc.laz"
+
+    reader = copc.FileReader("data/autzen-classified.copc.laz")
+
+    cfg = copc.LasConfig(reader.GetLasHeader(), reader.GetExtraByteVlr())
+    writer = copc.FileWriter(file_path, cfg)
+
+    root_page = writer.GetRootPage()
+
+    # TODO[Leo]: Find out why this test is so slow
```
Comment (CCInc) [inline test/writer_test.py:164]: todo

File: python/bindings.cpp [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+        .def("InvalidKey", &VoxelKey::InvalidKey)
+        .def("Bisect", &VoxelKey::Bisect)
+        .def("GetParent", &VoxelKey::GetParent)
+        .def("GetParents", &VoxelKey::GetParents, py::arg("include_current"))
+        .def("ChildOf", &VoxelKey::ChildOf, py::arg("parent_key"))
+        .def("__str__", &VoxelKey::ToString)
+        .def("__repr__", &VoxelKey::ToString);
+
+    py::class_<las::Point>(m, "Point")
+        .def(py::init<const uint8_t &, const uint16_t &>(), py::arg("point_format_id"), py::arg("num_extra_bytes") = 0)
```
Comment (CCInc) [inline python/bindings.cpp:45]: we can remove the consotructors?

File: example/example-writer.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+        points.AddPoint(point)
+    return points
+
+
+# In this example, we'll create our own file from scratch
+def NewFileExample():
+
+    # Create our new file with the specified format, scale, and offset
+    cfg = copc.LasConfig(8, copc.vector3(1, 1, 1), copc.vector3(0, 0, 0))
```
[the remark this answers is not in the document]
Comment (CCInc) [reply example/example-writer.py]: Or, even better, just use named args so it's clear what's going on :)

File: python/bindings.cpp [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+        .def_readwrite("offset", &las::LasHeader::offset)
+        .def_readwrite("header_size", &las::LasHeader::header_size)
+        .def_readwrite("point_format_id", &las::LasHeader::point_format_id)
+        .def_readwrite("point_record_length", &las::LasHeader::point_record_length)
+        .def_readwrite("point_count", &las::LasHeader::point_count);
+
+    py::class_<las::EbVlr>(m, "EbVlr").def(py::init<int>()).def_readwrite("items", &las::EbVlr::items);
+
+    py::class_<las::EbVlr::ebfield>(m, "EbField")
+        //    .def_readwrite("reserved",&las::EbVlr::ebfield::reserved)
```
[the remark this answers is not in the document]
Comment (CCInc) [reply python/bindings.cpp]: we decided to not do the eb in this PR

File: setup.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+        subprocess.check_call(
+            ["cmake", "--build", "."] + build_args, cwd=self.build_temp
+        )
+
+
+# The information here can also be placed in setup.cfg - better separation of
+# logic and declaration, and simpler if you include description/version in a file.
+setup(
+    name="copclib",
+    version="1.2",
```
[the remark this answers is not in the document]
Comment (CCInc) [reply setup.py:127]: feat-435

File: src/io/writer_internal.cpp [hunk 18]
```diff
@@ -84,7 +89,7 @@ void WriterInternal::WriteHeader(las::LasHeader &head14)
     lazVlr.header().write(out_stream_);
     lazVlr.write(out_stream_);
 
-    if (head14.ebCount())
+    if (head14.NumExtraBytes())
```
[the remark this answers is not in the document]
Comment (CCInc) [reply src/io/writer_internal.cpp:92]: just for the explicitness, I think it's better to make clear what non-boolean truth statements are doing

File: python/bindings.cpp (hunk 14 above)
Comment (CCInc) [reply python/bindings.cpp:45]: Or we can do this in another PR

**Review by CCInc — APPROVED** [review]

**Pull request merged**